daemon/config/config_test.go

```diff
@@ -13,41 +13,68 @@ import (
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/imdario/mergo"
 	"github.com/spf13/pflag"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/unicode"
 	"gotest.tools/v3/assert"
 	is "gotest.tools/v3/assert/cmp"
-	"gotest.tools/v3/fs"
 	"gotest.tools/v3/skip"
 )
 
+func makeConfigFile(t *testing.T, content string) string {
+	t.Helper()
+	name := filepath.Join(t.TempDir(), "daemon.json")
+	err := os.WriteFile(name, []byte(content), 0666)
+	assert.NilError(t, err)
+	return name
+}
+
 func TestDaemonConfigurationNotFound(t *testing.T) {
 	_, err := MergeDaemonConfigurations(&Config{}, nil, "/tmp/foo-bar-baz-docker")
 	assert.Check(t, os.IsNotExist(err), "got: %[1]T: %[1]v", err)
 }
 
 func TestDaemonBrokenConfiguration(t *testing.T) {
-	f, err := os.CreateTemp("", "docker-config-")
-	assert.NilError(t, err)
-
-	configFile := f.Name()
-	f.Write([]byte(`{"Debug": tru`))
-	f.Close()
+	configFile := makeConfigFile(t, `{"Debug": tru`)
 
-	_, err = MergeDaemonConfigurations(&Config{}, nil, configFile)
+	_, err := MergeDaemonConfigurations(&Config{}, nil, configFile)
 	assert.ErrorContains(t, err, `invalid character ' ' in literal true`)
 }
 
-// TestDaemonConfigurationWithBOM ensures that the UTF-8 byte order mark is ignored when reading the configuration file.
-func TestDaemonConfigurationWithBOM(t *testing.T) {
-	configFile := filepath.Join(t.TempDir(), "daemon.json")
+// TestDaemonConfigurationUnicodeVariations feeds various variations of Unicode into the JSON parser, ensuring that we
+// respect a BOM and otherwise default to UTF-8.
+func TestDaemonConfigurationUnicodeVariations(t *testing.T) {
+	jsonData := `{"debug": true}`
 
-	f, err := os.Create(configFile)
-	assert.NilError(t, err)
-
-	f.Write([]byte("\xef\xbb\xbf{\"debug\": true}"))
-	f.Close()
-
-	_, err = MergeDaemonConfigurations(&Config{}, nil, configFile)
-	assert.NilError(t, err)
+	testCases := []struct {
+		name     string
+		encoding encoding.Encoding
+	}{
+		{
+			name:     "UTF-8",
+			encoding: unicode.UTF8,
+		},
+		{
+			name:     "UTF-8 (with BOM)",
+			encoding: unicode.UTF8BOM,
+		},
+		{
+			name:     "UTF-16 (BE with BOM)",
+			encoding: unicode.UTF16(unicode.BigEndian, unicode.UseBOM),
+		},
+		{
+			name:     "UTF-16 (LE with BOM)",
+			encoding: unicode.UTF16(unicode.LittleEndian, unicode.UseBOM),
+		},
```

Comment:
I wonder if we're taking this a bit too far; would it be reasonable to tell that the file is expected to be in UTF-8? We have many JSON files, and don't handle them with the same assumptions.

(Als not a huge fan of introducing 3K+ lines of dependency, more so because there's been a lot of complexity in that dependency, and it had some vulnerabilities in the past).

Reply:
The daemon _writes_ many JSON files, and can therefore reasonably assume they will remain in the same encoding as written. The daemon config is a special case of a file which the user creates and the daemon only reads.

Reply:
Right, we should document what format we expect; I think it would be reasonable to document "must be UTF-8, and a BOM is allowed".

ISTR we strip the BOM for Dockerfiles as well, but don't do other conversions beyond that.

Reply:
I tend to agree with Seb here -- `TrimBytes` is really unobtrusive (fully stdlib!) and covers the most common cases of failure here, but it feels a little excessive to bring in such a large extra dependency just to enable users to not have to care what their editor does to the config file (and even `notepad` has gotten better about not doing this recently).

Reply:
The main reason for this/the issue I see is Windows Server; it simply is **hard** to write UTF-8 with no BOM on Windows Server 2019. Specifically, it:

* does not allow writing UTF-8 with no BOM in Notepad.exe; one must bail out to 'ANSI' and write UTF-8 escape sequences
* PowerShell does not support UTF-8 with no BOM except in [one Cmdlet](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-7.3#character-encoding-in-windows-powershell), `New-Item -Type File -Value`
* defaults to UTF-16LE or the system encoding in many places; when the system encoding is UTF-8, a BOM is often written

I think this is different than the JSON API or other interfaces to the Daemon as many of these tools are **expected ways to interact with the daemon**. Writing a Dockerfile is different as typically it is done with developer tools e.g. feature-rich text editors; the common tool to interact with `daemon.json` is PowerShell or Notepad, as the only text editor available out of the box on Windows.

We can certainly detect UTF-16 and yell at the user in a useful manner, which helps address the inscrutable errors I'm looking to improve here; however, that's still somewhat user-hostile as PowerShell's default encoding for `>`, `>>`, `Out-File`, and friends will be UTF-16.

I don't think extra dependencies is a huge concern here; we bring in `golang.org/x/text` and `golang.org/x/transform` already, and these are well-known, well-used, and well-audited libraries.

Reply:
I've posted a stdlib-only alternative at https://github.com/moby/moby/pull/44796; it's not quite as user friendly, and I still prefer this approach.

To add a little more signal: this change is driven by feedback/conversations with the Mirantis support org, as well as an examination of customer questions and support cases, and having to write a 500-word article on what tools can safely manipulate daemon.json on a barebones Windows Server install.

I think one way or another we should improve here, and the other PR gets us most of the way there; however, I still would like to reduce confusion and frustration by making the operating system standard tools "just work," more or less.